packages/checkout-storefront/src/sections/CheckoutForm/useCheckoutFormValidation.ts

```diff
@@ -0,0 +1,143 @@
+import { UsePaymentMethods } from "@/checkout-storefront/sections/PaymentSection";
+import { useAuthState } from "@saleor/sdk";
+import { useEffect } from "react";
+import { UseFormReturn } from "react-hook-form";
+import { ValidationError } from "yup";
+
+interface UseCheckoutFormValidation
+  extends Pick<UsePaymentMethods, "isValidProviderSelected">,
+    UseFormReturn<CheckoutFormData> {
+  schema: { validateSyncAt: (key: keyof CheckoutFormData, data: CheckoutFormData) => void };
+}
+
+export const useCheckoutFormValidation = ({
+  isValidProviderSelected,
+  watch,
+  setValue,
+  getValues,
+  schema,
+}: UseCheckoutFormValidation) => {
+  const formatMessage = useFormattedMessages();
+  const { checkout } = useCheckout();
+  const { shippingAddress, billingAddress } = checkout;
+  const { showCustomErrors } = useAlerts("checkoutFinalize");
+  const { authenticated } = useAuthState();
+
+  const {
+    hasAllRequiredFields: shippingHasAllRequiredFields,
+    getMissingFieldsFromAddress: getMissingFieldsFromShipping,
+  } = useAddressFormUtils(shippingAddress?.country?.code as CountryCode);
+
+  const {
+    hasAllRequiredFields: billingHasAllRequiredFields,
+    getMissingFieldsFromAddress: getMissingFieldsFromBilling,
+  } = useAddressFormUtils(billingAddress?.country?.code as CountryCode);
+
+  const getShippingMissingFieldsErrorMessage = () =>
+    getAddressMissingFieldsErrorMessage(
+      "missingFieldsInShippingAddress",
+      getMissingFieldsFromShipping(shippingAddress)
+    );
+
+  const getBillingMissingFieldsErrorMessage = () =>
+    getAddressMissingFieldsErrorMessage(
+      "missingFieldsInBillingAddress",
+      getMissingFieldsFromBilling(billingAddress)
+    );
+
+  const getAddressMissingFieldsErrorMessage = (messageKey: MessageKey, fields: AddressField[]) =>
+    `${formatMessage(messageKey)}: ${fields
+      .map((field) => formatMessage(field as MessageKey))
+      .join(", ")}`;
+
+  const ensureValidCheckout = (): boolean => {
+    let isValid = true;
+    setValue("validating", true);
+    const formData = getValues();
+    const { createAccount } = formData;
+
+    try {
+      const isLoggedIn = authenticated && checkout?.email;
+
+      if (!isLoggedIn) {
+        schema.validateSyncAt("email", formData);
+      }
+    } catch (e) {
+      const { path, type } = e as ValidationError;
+      showCustomErrors([
+        { field: path as string, code: type === "email" ? "invalid" : (type as ErrorCode) },
+      ]);
+      isValid = false;
+    }
+
+    if (createAccount) {
+      try {
+        schema.validateSyncAt("password", formData);
+      } catch ({ path, type }) {
+        showCustomErrors([{ field: path as string, code: type as ErrorCode }]);
+        isValid = false;
+      }
+    }
+
+    if (!checkout.shippingAddress) {
+      showCustomErrors([{ field: "shippingAddress", code: "required" }]);
+    }
+
+    if (!shippingHasAllRequiredFields(checkout.shippingAddress)) {
```

Comment:
Looks like we could use some form-splitting at this point, to make `shippingAddress` and `billingAddress` kind of independent forms with their own validation.